specs/experimental/custom-gas-token.md

```diff
@@ -0,0 +1,306 @@
+- `sourceHash`: TODO
+
+### L2CrossDomainMessenger Deployment
+
+- `from`: `0x4210000000000000000000000000000000000003`
+- `to`: `null`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `375,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+### L2StandardBridge Deployment
+
+- `from`: `0x4210000000000000000000000000000000000004`
+- `to`: `null`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `375,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+### L1Block Proxy Update
+
+- `from`: `0x0000000000000000000000000000000000000000`
+- `to`: `0x4200000000000000000000000000000000000015`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `50,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+### L2CrossDomainMessenger Proxy Update
+
+- `from`: `0x0000000000000000000000000000000000000000`
+- `to`: `0x4200000000000000000000000000000000000007`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `50,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+### L2StandardBridge Proxy Update
+
+- `from`: `0x0000000000000000000000000000000000000000`
+- `to`: `0x4200000000000000000000000000000000000010`
+- `mint`: `0`
+- `value`: `0`
+- `gasLimit`: `50,000`
+- `data`: TODO
+- `sourceHash`: TODO
+
+## Selection of `ETHER_TOKEN_ADDRESS`
+
+It was decided to use `address(0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE)` to represent `ether` to push ecosystem standardization
+efforts around using this address to represent `ether` in [DeFi protocols](https://twitter.com/transmissions11/status/1775525755860169010)
+or [APIs](https://github.com/ethereum/execution-apis/pull/484).
+
+## Security Considerations
+
+### OptimismPortal Token Allowance
+
+The `OptimismPortal` makes calls on behalf of users. It is therefore unsafe to be able to call the address of the custom gas
+token itself from the `OptimismPortal` because it would be a simple way to `approve` an attacker's balance and steal the
+entire ERC20 token balance of the `OptimismPortal`.
+
+### Decimal Scaling
+
+A library for scaling `uint256` representations of floating point numbers is used on both ERC20 token deposits and withdrawals.
+This ensures that from the point of view of the L2 system, it always appears as if the number of tokens that were deposited
+are the same as the native balance. Tokens with less than or 18 decimals are supported, meaning that there is no potential for
+loss of precision due to sending in too small of a unit.
+
+### Interoperability Support
+
+Interop is supported between chains that use a custom gas token. The token address and the number of decimals are legible on
+chain. In the future we may add the ability to poke a chain such that it emits an event that includes the custom gas token address
+and its number of decimals to easily be able to introspect on the native asset of another chain.
+
+### Wrapped Ether
+
+The `WETH9` predeploy at `0x4200000000000000000000000000000000000006` represents wrapped native asset and not wrapped `ether`.
+Portable and fungible `ether` across different domains is left for a future project.
```

Comment:
I agree this is the simplest approach, but:
1. Let's add rationale for this decision
2. We should update WETH9 to have a name of `"string.concat("Wrapped ", L1Block.gasPayingTokenName())` and similarly for symbol. This will require name/symbol to be stored in L1Block. Otherwise I think this will cause a lot of confusion for users because all block explorers, wallets, etc. will say they have "Wrapped Ether" when it's really a different token. A static name of "Wrapped Gas Paying Token" is still an improvement, and a smaller diff to implement, but still might be confusing for users who are using AA wallets and don't know what gas is. WETH9 is not proxied, but I think a version of this change is needed regardless.

Reply:
People will certainly be unhappy if we do not do this. Previously we didn't want to change the weth implementation due to the desire to just use the same bytecode. I decided on not including this to cut scope. Having some product insight here would be useful cc @K-Ho.  I am open to this change but it just adds extra scope

Reply:
I am tracking this in https://github.com/ethereum-optimism/protocol-quest/issues/222#issuecomment-2041776976